src/SFA.DAS.Roatp.CourseManagement.Web/Controllers/AddAStandard/AddLocationController.cs

```diff
@@ -0,0 +1,105 @@
+        private readonly IMediator _mediator;
+        private readonly ISessionService _sessionService;
+
+        public AddLocationController(IMediator mediator, ISessionService sessionService, ILogger<AddLocationController> logger)
+        {
+
+            _mediator = mediator;
+            _sessionService = sessionService;
+            _logger = logger;
+        }
+
+        [Route("{ukprn}/standards/{larsCode}/locations/add-new", Name = RouteNames.GetNewStandardAddProviderCourseLocation)]
+        [HttpGet]
+        public async Task<IActionResult> SelectAProviderlocation(int larsCode)
+        {
+            var model = await GetModel(larsCode);
+            if (model == null) return RedirectToRouteWithUkprn(RouteNames.GetAddStandardSelectStandard);
+            return View(ViewPath, model);
+        }
+
+        [Route("{ukprn}/standards/{larscode}/locations/add-new", Name = RouteNames.PostNewStandardAddProviderCourseLocation)]
+        [HttpPost]
+        public async Task<IActionResult> SubmitAProviderlocation(CourseLocationAddSubmitModel submitModel, int larsCode)
+        {
+            var model = await GetModel(larsCode);
+            if (model == null) return RedirectToRouteWithUkprn(RouteNames.GetAddStandardSelectStandard);
+            if (!ModelState.IsValid)
+            {
+                return View(ViewPath, model);
+            }
+   
+            var sessionModel = _sessionService.Get<StandardSessionModel>(Ukprn.ToString());
+            if (sessionModel.CourseLocations == null)
+                sessionModel.CourseLocations = new List<CourseLocationModel>() ;
+
+           sessionModel.CourseLocations.Add(new CourseLocationModel
+           {
+               LocationType = LocationType.Provider,
+               ProviderLocationId = Guid.Parse(submitModel.TrainingVenueNavigationId),
+               LocationName = model.TrainingVenues.First(x=>x.Value==submitModel.TrainingVenueNavigationId).Text,
+               DeliveryMethod = new DeliveryMethodModel
+               {
+                   HasBlockReleaseDeliveryOption = submitModel.HasBlockReleaseDeliveryOption,
+                   HasDayReleaseDeliveryOption = submitModel.HasDayReleaseDeliveryOption
+               }
+           });
+
+           _sessionService.Set(sessionModel, Ukprn.ToString());
+
+            return RedirectToRoute(RouteNames.GetNewStandardViewTrainingLocationOptions, new { ukprn = Ukprn });
+        }
+
+        private async Task<CourseLocationAddViewModel> GetModel(int larsCode)
+        {
+            _logger.LogInformation("Getting provider course locations for ukprn {ukprn} for larsCode {larsCode}", Ukprn, larsCode);
+
+            var result = await _mediator.Send(new GetAllProviderLocationsQuery(Ukprn));
+            var allProviderLocations = result.ProviderLocations;
+
+            var availableProviderLocations = new List<ProviderLocation>();
+
+            var sessionModel = _sessionService.Get<StandardSessionModel>(Ukprn.ToString());
```

Comment:
Move the session model check in the controller action and follow the pattern of using escape route. Pass the model in to this method as parameter. I know it is repeating, but I would prefer a tiny repeat to keep it consistent.